### ISSUE TITLE
Trace MPI_ functions in AMPI

### ISSUE THREAD

**Issue by stwhite91** [UIUC-PPL/charm#1088]
*Original issue: https://charm.cs.illinois.edu/redmine/issues/1088*

---
If built with tracing enabled (or perhaps as another option), AMPI should insert user events to all MPI_ routines to mark their beginning and end. This could potentially be added to the functionality of the AMPIAPI("MPI_Function_name") macros at the start of every MPI_ function.

**Comment by stwhite91** [comment]
*Original date: 2016-06-29 08:04:19*

---
Discussing this with Ronak, realized we should use system events rather than user events like Charm does for entry methods, so that these are separate from user events. Since we have deemed PMPI_ support impractical, this should be bumped up in priority.

We should be able to change AMPIAPI to include tracing stuff at the beginning and end of each AMPI_ routine. AMPIAPI already takes the function name, so we shouldn't even have to change anything other than the definition of AMPIAPI itself.

To do:
1. Add a routine that is called during AMPI's startup process (if tracing is enabled) to register every AMPI_ function with the tracing framework.
2. Add a call to start the trace for a particular function to TCharmAPIRoutine's constructor.
3. Add a call to stop the trace for a particular function to TCharmAPIRoutine's destructor.

We can do this with userBracketedEvents, but viewing them in Projections is not very helpful since the events span blocking events: for example, if we run with virtualization and trace a call to MPI_Barrier, we just see one event from the first rank on a PE to reach the barrier until the last rank on that PE exits that barrier.

Instead, we may want to insert calls to stop tracing whenever we block inside AMPI, so that tracing is split phase for such routines.

Basically, look at some AMPI Projections traces in Timeline view and see how we can improve them incrementally.

**Comment by matthiasdiener** [comment]
*Original date: 2017-04-24 18:38:23*

---
WIP patch here: ~~https://charm.cs.illinois.edu/gerrit/2426~~ https://github.com/UIUC-PPL/charm/commit/3026ae0aaff6749169e621803bb1bee2dfa51c08

**Comment by stwhite91** [comment]
*Original date: 2017-05-03 01:53:32*

---
2 things to follow up on:
1. Use unordered_map instead of map (use tr1::unordered_map if CMK_USING_XLC before 6.8.0)
2. From your progress report, why are the bracketed user events overlapping in time when it seems like they shouldn't be?

**Comment by matthiasdiener** [comment]
*Original date: 2017-05-03 15:27:39*

---
Another thing to followup:
- MPI_Finalize() does not get traced, because the thread gets only suspended, and the TCHARM_API_TRACE desctrutor never gets executed.

EDIT: Fixed here: ~~https://charm.cs.illinois.edu/gerrit/#/c/2550/~~ https://github.com/UIUC-PPL/charm/commit/4aa8a4429aeae8fe038632dc37867bb042eac447

**Comment by stwhite91** [comment]
*Original date: 2017-05-03 16:42:28*

---
Also, clean up heap memory allocated for the funcmap. I made a half-baked attempt at that here but it has issues noted here: https://charm.cs.illinois.edu/gerrit/#/c/2463/ 

To do this properly, the easiest thing might be to have the funcmap be owned by a Node Group, and during the TCharm exit sequence every thread contributes to a reduction over that Node Group, witha  callback that broadcasts to the node group and deletes that memory, before continuing on to call CkExit().

**Comment by matthiasdiener** [comment]
*Original date: 2017-05-07 00:03:44*

---
For the second issue (overlapping events), I created a new bug (#1551).

**Comment by stwhite91** [comment]
*Original date: 2017-05-21 20:03:28*

---
Change std::map to std::unordered_map: ~~https://charm.cs.illinois.edu/gerrit/#/c/2545/~~ https://github.com/UIUC-PPL/charm/commit/d7db3cd67fbf17ffe158720d9f6206fedd02dcf4
